LN review 20210528

LawrenceLoz opened the pull request:
Most changes are around the guidance for unlocked packages.

 - Added suggestions for managing objects within domain packages
 - Provided more detail on components which should
 - Clarified that profiles can't be included in unlocked packages
 - Removed the explicit suggestion to consider not packaging components (e.g. layouts) which have many dependencies. Feel free to revert this if you disagree, but my view is it's better to have one or more packages with dependencies on many unlocked packages to include these global components and benefit from packaging features such as managed metadata deprecation
 - Added other examples of metadata which we might not want to package

Also added examples of OSS frameworks to help with feature toggling and a few extra points in this section

azlam-abdulsalam commented:
Thanks @LawrenceLoz  for your considerable amount of time into this. This looks really good

rakr commented on development-practices/unlocked-packages.md:
```diff
@@ -34,13 +34,19 @@ The following sections deal with items that are particular to DX@Scale or more e
-* **Workflows and ProcessBuilders** need to be placed in the same unlocked package that contains the parent object definition. This is a restriction of this particular metadata component, In these scenarios rather than building  automation using workflows and process builders, it is better to use apex or flow
+* Don’t package the metadata merely because it is supported by Unlocked Packaging. Some components (e.g. profiles) will need to be managed outside of unlocked packages and it may be preferable to deploy other related metadata using the same mechanism. In other cases, metadata like non-critical reports, email templates etc, may be better to manage outside of source control entirely to allow these  ...
+* **Objects** should be "owned", where possible, by one domain specific package which includes the majority of fields, validation rules and other metadata related to this object. Other metadata acting on the object should be scoped where possible to operate within the confines of the package. Examples include:
+  * **Permission Sets**, which should be granular and specific to an object or package
```
We run the risk of having very SObject related permission sets, agreed that the permission sets must be granular at the function level, for example: **ability to create customers** which leads to the creation of a permission set that enables the creation/modification of account and contact records or **ability to administrate customers** which would lead to another permission set that would enable delete/viewAll permissions on the same SObject.
Probably also worth mentioning that the permission sets should be composable.

LawrenceLoz replied:
Yes good point. I've recommended object specific in orgs I've worked on partly because it's quite a clear way of ensuring granularity but you're right the main point is the granularity and the object-specific aspect is pretty arbitrary. I've updated this point with your examples and a few more details. And 100% on the composable aspect, this is really crucial, I've added a point around this too

rakr commented on development-practices/unlocked-packages.md:
```diff
@@ -34,13 +34,19 @@ The following sections deal with items that are particular to DX@Scale or more e
+  * **Apps**, which include tabs from multiple packages
+  * **Layouts**, which can include related lists, buttons and fields referencing metadata outside of the object's package
+  * **Flexipages**, which may include components, actions and lists from across packages
+* **Profiles** cannot be included in unlocked packages and need[ special attention](https://docs.dxatscale.io/scm/managing-profiles).
```
:)

rakr replied:
Me happy!

rakr commented on scm/branching-model/feature-toggling.md:
```diff
@@ -1,6 +1,8 @@
 # Feature Toggling
 
-Following considerations need to be taken care when a feature toggling approach is considered
+Feature Toggles allow new components and changes to be deloyed in an inactive state and enabled at a time when stakeholders or technical dependencies are ready.
```
typo in the word "deployed"

LawrenceLoz replied:
Thanks updated

rakr commented on scm/branching-model/feature-toggling.md:
```diff
@@ -1,6 +1,8 @@
-Following considerations need to be taken care when a feature toggling approach is considered
+Feature Toggles allow new components and changes to be deloyed in an inactive state and enabled at a time when stakeholders or technical dependencies are ready.
+
+This pattern can decouple the risk and logistics of the deployment itself from the business considerations involved in adopting a new feature, and helps to support the practice of small and frequent deployments.
```
Probably worth mentioning that the first feature toggling mechanism that can be implemented is "do we deploy the feature or not". Also, feature toggles are usually transient, we deploy a new feature, we test it then the feature stays. Beyond enabling feature toggling, removing the toggles after a while is something that need to be worked on (note to myself here)

LawrenceLoz replied:
Yup, I've included a bullet point at the end about not deploying. Also there was a bullet point there already about removing toggles which is a good reminder

rakr commented on scm/branching-model/feature-toggling.md:
```diff
@@ -10,23 +12,23 @@ Workflow, Process Builders and Visual Workflows \(branching condition that check
 
-* Using a Trigger Framework which allows for deactivation of any given trigger handler via a metadata type
-* Use a framework to read values from a custom metadata object to determine whether to execute code
+* Use a Trigger Framework which allows for deactivation of any given trigger handler via a Custom Metadata Type or Custom Permission. [Apex Trigger Actions](https://github.com/mitchspano/apex-trigger-actions-framework) and [Nebula Triggers](https://bitbucket.org/nebulaconsulting/nebula-core/src/master/) are examples of open source framerworks which provide this capability
```
I am not sure if the ability to enable or disable a trigger handler is desirable feature. Probably worth discussing the use cases before promoting this.
Also typo in the word "frameworks"

azlam-abdulsalam replied:
Thats probably me..as part of the original content, will wait for your both decision and change accordingly

LawrenceLoz replied:
My view would be this can be a useful and straightforward way to toggle for some circumstances - for example a trigger handler which calls a service to make a REST request to a new external system and update the record with the result can be inactive until the system is up.

Worth mentioning the meaning of "handler" under each of these frameworks would be a discrete apex class doing one aspect of trigger processing, so disabling one class still allows other logic from the trigger to run

rakr reviewed (commented):
Great stuff @LawrenceLoz, I've added few comments in the PR, can you please review and let me know what you think

LawrenceLoz commented on development-practices/unlocked-packages.md:
```diff
@@ -34,13 +34,19 @@ The following sections deal with items that are particular to DX@Scale or more e
-  * Permission Sets
-  * Layouts    
-* **Workflows and ProcessBuilders** need to be placed in the same unlocked package that contains the parent object definition. This is a restriction of this particular metadata component, In these scenarios rather than building  automation using workflows and process builders, it is better to use apex or flow
+* Don’t package the metadata merely because it is supported by Unlocked Packaging. Some components (e.g. profiles) will need to be managed outside of unlocked packages and it may be preferable to deploy other related metadata using the same mechanism. In other cases, metadata like non-critical reports, email templates etc, may be better to manage outside of source control entirely to allow these  ...
```
Agree with the points around reports - I've added this in (with this and other points it feels helpful to separate into What to Include / What Not to Include sections)

For the profiles my perspective is that in some environments the version in production should be controlled more carefully given the potential risks around direct changes and the importance of having an up to date version of what's in production, and that autonomy should be given to the admins for to create permission sets to extend very lightweight profiles. I've described these two alternatives in the profiles page and reasons you might pick one over the other - appreciate you might prefer to recommend a single strategy and happy to change this to include just the org-dependent packages approach if you prefer?

LawrenceLoz commented:
Thanks for your considered feedback @rakr - some excellent points! I've reworked the changes in response to these and added a couple of questions/thoughts above. Let me know if anything else you think should be tweaked
